File: Back-End/ml/occupation_ml.py
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense
import datetime
import json

# Load the JSON file
with open('data.json', 'r') as f:
    data = json.load(f)

time_data = data['time']
occupation_data = data['occupation']

# Prepare data
df = pd.DataFrame({'time': time_data, 'occupation': occupation_data})
df['time'] = pd.to_datetime(df['time'])
df['day_of_week'] = df['time'].dt.dayofweek  # 0 for Monday, 6 for Sunday
df['hour'] = df['time'].dt.hour
df['minute'] = df['time'].dt.minute

encoder = OneHotEncoder(sparse_output=False)
day_of_week_encoded = encoder.fit_transform(df[['day_of_week']])

time_features = df[['hour', 'minute']].values
features_combined = np.concatenate([day_of_week_encoded, time_features], axis=1)

feature_scaler = MinMaxScaler(feature_range=(0, 1))
features_combined[:, -2:] = feature_scaler.fit_transform(features_combined[:, -2:])

target = df['occupation'].values

target_scaler = MinMaxScaler(feature_range=(0, 1))
target_scaled = target_scaler.fit_transform(target.reshape(-1, 1))

sequence_length = len(time_data)
X, y = [], []
for i in range(sequence_length, len(features_combined)):
    X.append(features_combined[i-sequence_length:i])
    y.append(target_scaled[i])

X, y = np.array(X), np.array(y)

# Define and train the model
model = Sequential()
model.add(LSTM(50, return_sequences=True, input_shape=(X.shape[1], X.shape[2])))
model.add(LSTM(50, return_sequences=False))
model.add(Dense(25))
model.add(Dense(1))

model.compile(optimizer='adam', loss='mean_squared_error')
model.fit(X, y, batch_size=1, epochs=10)

# Function to predict occupation for each hour of the current day
def predict_for_day():
    now = datetime.datetime.now()
    weekday_number = now.weekday()

    predictions = []

    for hour in range(24):
        # Prepare input for each hour
        new_day = np.array([[weekday_number]])
        new_day_encoded = encoder.transform(new_day)

        new_time_features = feature_scaler.transform([[hour, 0]])  # Set minute to 0 for hourly prediction

        new_features_combined = np.concatenate([new_day_encoded, new_time_features], axis=1)
        new_sequence = np.tile(new_features_combined, (sequence_length, 1))
        new_sequence = new_sequence.reshape(1, sequence_length, new_features_combined.shape[1])

        # Predict occupation
        predicted_occupation_scaled = model.predict(new_sequence)
        predicted_occupation = target_scaler.inverse_transform(predicted_occupation_scaled)

        predictions.append(predicted_occupation[0][0])

    return predictions

# Get predictions for the day
predictions = predict_for_day()

# Save predictions to a .js file
output_data = {
    "predictions": predictions
}

with open('', 'w') as f:
    f.write('const predictions = ' + json.dumps(output_data, indent=4) + ';\n')
    f.write('module.exports = predictions;\n')

print("Predictions have been saved to predictions.js")

File: Back-End/ml/occupation_ml.js
import { getData } from "../app";
let data;

// Event listener for user click
document.addEventListener("UserClickedOnList", async (event) => {
  const params = new URLSearchParams(window.location.search);
  const place_id = params.get("id");
  const metadata = await getData(place_id);
  data = {
    time: metadata['time'],
    occupation: metadata['occupation']
  } 
  || 
  // Example data
  {
    time: [
      '2000-01-03T10:00:00.000+00:00', // Monday
      '2000-01-04T10:00:00.000+00:00', // Tuesday
      '2000-01-05T10:00:00.000+00:00', // Wednesday
      '2000-01-06T10:00:00.000+00:00', // Thursday
      '2000-01-07T10:00:00.000+00:00', // Friday
      '2000-01-08T10:00:00.000+00:00', // Saturday
      '2000-01-09T10:00:00.000+00:00', // Sunday
    ],
    occupation: [80, 75, 70, 85, 90, 60, 65]
  };
});

const tf = require('@tensorflow/tfjs-node');
const moment = require('moment');
const { MinMaxScaler } = require('scikitjs');
const scikit = require('scikitjs');
scikit.setBackend(tf);

// Preprocessing
const timesMoment = data.time.map(time => moment(time));
const daysOfWeek = timesMoment.map(time => time.day()); // 0 for Sunday, 6 for Saturday
const hours = timesMoment.map(time => time.hour());
const minutes = timesMoment.map(time => time.minute());

// One-hot encode day of the week
const dayOfWeekEncoded = tf.oneHot(tf.tensor1d(daysOfWeek, 'int32'), 7).arraySync();

// Combine features
let featuresCombined = dayOfWeekEncoded.map((day, i) => [...day, hours[i], minutes[i]]);

// Scale features
const featureScaler = new MinMaxScaler({ featureRange: [0, 1] });
featuresCombined = featureScaler.fitTransform(featuresCombined);

// Scale target
const targetScaler = new MinMaxScaler({ featureRange: [0, 1] });
const targetScaled = targetScaler.fitTransform(data.occupation.map(o => [o]));

// Prepare sequences
const sequenceLength = data.time.length;
const X = [];
const y = [];
for (let i = sequenceLength; i < featuresCombined.length; i++) {
  X.push(featuresCombined.slice(i - sequenceLength, i));
  y.push(targetScaled[i]);
}

const XTensor = tf.tensor3d(X);
const yTensor = tf.tensor2d(y);

// Build model
const model = tf.sequential();
model.add(tf.layers.lstm({ units: 50, returnSequences: true, inputShape: [XTensor.shape[1], XTensor.shape[2]] }));
model.add(tf.layers.lstm({ units: 50, returnSequences: false }));
model.add(tf.layers.dense({ units: 25 }));
model.add(tf.layers.dense({ units: 1 }));

model.compile({ optimizer: 'adam', loss: 'meanSquaredError' });

// Train model
(async () => {
  await model.fit(XTensor, yTensor, { batchSize: 1, epochs: 10 });

  // Prediction function for the current weekday occupation (full 24-hour period)
  const predictRealTimeWeekdayOccupation = async () => {
    const now = moment();
    const weekdayNumber = now.day(); // Get the current weekday (0-6)

    const predictedOccupations = [];

    for (let hour = 0; hour < 24; hour++) {
      const minute = 0; // Predict at the start of each hour

      // One-hot encode the current weekday
      const dayEncoded = tf.oneHot(tf.tensor1d([weekdayNumber], 'int32'), 7).arraySync()[0];
      
      // Scale hour and minute features
      const timeFeaturesScaled = featureScaler.transform([[hour, minute]])[0];

      // Combine day and time features
      const featuresCombined = [...dayEncoded, ...timeFeaturesScaled];
      
      // Create a sequence for prediction
      const sequence = Array(sequenceLength).fill(featuresCombined);

      // Convert to tensor for prediction
      const sequenceTensor = tf.tensor3d([sequence]);

      // Predict the occupation
      const predictedOccupationScaled = model.predict(sequenceTensor);
      const predictedOccupation = targetScaler.inverseTransform(predictedOccupationScaled.arraySync());

      predictedOccupations.push(predictedOccupation[0][0]);
    }

    console.log('Predicted Occupations for Each Hour Today:', predictedOccupations);
  };

  // Predict the occupation for the current weekday
  await predictRealTimeWeekdayOccupation();
})();
